text/0004-remove-unnecessary-allocator-usage.md

```diff
@@ -0,0 +1,328 @@
+
+The encoding of the response code is also modified compared to its version 1 counterpart and each response code now encodes to 4 little endian bytes as described below:
+
+- 100-999: the request has finished with the given HTTP status code.
+- -1 if the deadline was reached.
+- -2 if there was an I/O error while processing the request.
+- -3 if the identifier of the request is invalid.
+
+The buffer passed to `out` must always have a size of `4 * n` where `n` is the number of elements in the `ids`.
+
+```wat
+(func $ext_offchain_http_response_header_name_version_1
+    (param $request_id i32) (param $header_index i32) (param $out i64) (result i64))
+(func $ext_offchain_http_response_header_value_version_1
+    (param $request_id i32) (param $header_index i32) (param $out i64) (result i64))
+```
+
+These functions supercede the `ext_offchain_http_response_headers_version_1` host function.
+
+Contrary to `ext_offchain_http_response_headers_version_1`, only one header indicated by `header_index` can be read at a time. Instead of calling `ext_offchain_http_response_headers_version_1` once, the runtime should call `ext_offchain_http_response_header_name_version_1` and `ext_offchain_http_response_header_value_version_1` multiple times with an increasing `header_index`, until a value of `-1` is returned.
+
+These functions accept an `out` parameter containing [a pointer-size](https://spec.polkadot.network/chap-host-api#defn-runtime-pointer-size) to the memory location where the header name or value should be written. The runtime execution stops with an error if `out` is outside of the range of the memory of the virtual machine, even if the function wouldn't write anything to `out`.
+
+These functions return the size, in bytes, of the header name or header value. If request doesn't exist or is in an invalid state (as documented for `ext_offchain_http_response_headers_version_1`) or the `header_index` is out of range, a value of `-1` is returned. Given that the host must never write more bytes than the size of the buffer in `out`, and that the size of this buffer is expressed as a 32 bits number, a 64bits value of `-1` is not ambiguous.
+
+If the buffer in `out` is too small to fit the entire header name of value, only the bytes that fit are written and the rest are discarded.
+
+```wat
+(func $ext_offchain_submit_transaction_version_2
+    (param $data i64) (return i32))
+(func $ext_offchain_http_request_add_header_version_2
+    (param $request_id i32) (param $name i64) (param $value i64) (result i32))
+```
+
+The behaviour of these functions is identical to their version 1 counterpart. Instead of allocating a buffer, writing `1` or `0` in it, and returning a pointer to it, the version 2 of these functions simply return 1 or 0.
```

Comment:
```suggestion
The behaviour of these functions is identical to their version 1 counterpart. Instead of allocating a buffer, writing `1` or `0` in it, and returning a pointer to it, the version 2 of these functions simply return `1` or `0`. While `0` means that the operation was successful. 
```

Reply:
Changed